tools/tesseratos/src/tesseratos/console/plugin.cpp

```diff
@@ -0,0 +1,270 @@
+{
+    switch (level)
+    {
+    case cubos::core::Logger::Level::Trace:
+        return "Trace";
+    case cubos::core::Logger::Level::Debug:
+        return "Debug";
+    case cubos::core::Logger::Level::Info:
+        return "Info";
+    case cubos::core::Logger::Level::Warn:
+        return "Warn";
+    case cubos::core::Logger::Level::Error:
+        return "Error";
+    case cubos::core::Logger::Level::Critical:
+        return "Critical";
+    default:
+        return "Unknown";
+    }
+}
+
+// Function to convert log level to a color
+static ImVec4 levelToColor(cubos::core::Logger::Level level)
+{
+    switch (level)
+    {
+    case cubos::core::Logger::Level::Trace:
+        return {1.0F, 1.0F, 1.0F, 1.0F}; // White
+    case cubos::core::Logger::Level::Debug:
+        return {0.0F, 1.0F, 1.0F, 1.0F}; // Cyan
+    case cubos::core::Logger::Level::Info:
+        return {0.0F, 1.0F, 0.0F, 1.0F}; // Green
+    case cubos::core::Logger::Level::Warn:
+        return {1.0F, 1.0F, 0.0F, 1.0F}; // Yellow
+    case cubos::core::Logger::Level::Error:
+        return {1.0F, 0.0F, 0.0F, 1.0F}; // Red
+    case cubos::core::Logger::Level::Critical:
+        return {1.0F, 0.0F, 0.0F, 0.5F}; // Bright Red
+    default:
+        return {1.0F, 1.0F, 1.0F, 1.0F}; // Default to
+    }
+}
+
+void tesseratos::consolePlugin(Cubos& cubos)
+{
+    cubos.depends(cubos::engine::imguiPlugin);
+    cubos.depends(toolboxPlugin);
+    cubos.resource<State>();
+
+    cubos.system("show Logger UI").tagged(cubos::engine::imguiTag).call([](Toolbox& toolbox, State& state) {
+        if (!toolbox.isOpen("Console"))
+        {
+            return;
+        }
+
+        ImGui::SetNextWindowSize(ImVec2(930, 310), ImGuiCond_FirstUseEver);
+        ImGui::SetNextWindowPos(ImVec2(10, 10), ImGuiCond_FirstUseEver);
+        ImGui::Begin("Console");
+
+        ImGui::Columns(2, "ConsoleLayout", true);
+
+        // Log Scrollable Window
+        ImGui::SetColumnWidth(0, ImGui::GetWindowWidth() * 0.8F);
+
+        std::size_t cursor = state.prevCursor;
+        cubos::core::Logger::Entry entry;
+        cubos::core::Logger::Entry previousEntry;
+
+        // Counters for each log level
+        std::unordered_map<cubos::core::Logger::Level, int> logCounts = {
+            {cubos::core::Logger::Level::Trace, 0}, {cubos::core::Logger::Level::Debug, 0},
+            {cubos::core::Logger::Level::Info, 0},  {cubos::core::Logger::Level::Warn, 0},
+            {cubos::core::Logger::Level::Error, 0}, {cubos::core::Logger::Level::Critical, 0}};
+
+        ImGui::BeginChild("Log", ImVec2(0, -ImGui::GetFrameHeightWithSpacing()));
+
+        state.uiEntries.clear();
+
+        while (cubos::core::Logger::read(cursor, entry))
+        {
+            state.uiEntries.push_back(entry);
+        }
+
+        if (state.clearEnabled)
+        {
+            state.clearEnabled = false;
+            state.prevCursor = cursor;
+        }
```

Comment:
```suggestion
        cubos::core::Logger::Entry entry;
        cubos::core::Logger::Entry previousEntry;

        // Counters for each log level
        std::unordered_map<cubos::core::Logger::Level, int> logCounts = {
            {cubos::core::Logger::Level::Trace, 0}, {cubos::core::Logger::Level::Debug, 0},
            {cubos::core::Logger::Level::Info, 0},  {cubos::core::Logger::Level::Warn, 0},
            {cubos::core::Logger::Level::Error, 0}, {cubos::core::Logger::Level::Critical, 0}};

        ImGui::BeginChild("Log", ImVec2(0, -ImGui::GetFrameHeightWithSpacing()));

        while (cubos::core::Logger::read(state.prevCursor, entry))
        {
            state.uiEntries.push_back(entry);
        }

        if (state.clearEnabled)
        {
            state.clearEnabled = false;
            state.uiEntries.clear();
        }
```

You want to store the `cursor` written to by `Logger::read`. This way, on each call to this system (on each frame), you only push_back new log entries to `uiEntries`, and thus you don't need to clear it every frame! This is way more efficient.
When you want to forget 'past history', you simply clear the `uiEntries` vector.